### PR TITLE
Disallow metadata downgrades

### PR DESCRIPTION
**Pull request by mwleeds**
Like https://github.com/flatpak/flatpak/commit/ed1d7eacf47534fc6aae6571b7d41bca3951d0e9 this is more of a hack than a complete fix, but I think it's more desirable.

### CODE REVIEW

File: common/flatpak-error.h [hunk 1]
```diff
@@ -42,6 +42,8 @@ G_BEGIN_DECLS
  * @FLATPAK_ERROR_NEED_NEW_FLATPAK: The App/Runtime needs a more recent version of flatpak.
  * @FLATPAK_ERROR_REMOTE_NOT_FOUND: The specified remote was not found.
  * @FLATPAK_ERROR_RUNTIME_NOT_FOUND: An runtime needed for the app was not found.
+ * @FLATPAK_ERROR_DOWNGRADE: The pulled commit is a downgrade, and a downgrade wasn't
+ *                           specifically allowed.
```
Comment (pwithnall) [inline common/flatpak-error.h]: Add `(Since: 1.0)` (or whatever version this is going to be released in). i.e. `The pulled commit … specifically allowed. (Since: 1.0)`

Comment (mwleeds) [reply common/flatpak-error.h]: Pushed a fixup

**Comment by mwleeds** [comment]
Just figured out this doesn't work properly when the system-helper does the metadata pull, because [this line](https://github.com/flatpak/flatpak/blob/745287e48fade2a6f3dd5368ebd280a8c3039f3c/system-helper/flatpak-system-helper.c#L358) covers up the error type. Not sure what to do about that other than search the error message for the string "downgrade" which seems terrible.

**Comment by alexlarsson** [comment]
We should be able to pass a FLATPAK_ERROR_* across the dbus error, because we register it with dbus using g_dbus_error_register_error_domain() in flatpak_error_quark(). It might need some decoding with g_dbus_error_strip_remote_error() tho (i never quite figured out when that is needed).

**Comment by ramcq** [comment]
> We should be able to pass a FLATPAK_ERROR_* across the dbus error, because we register it with dbus using g_dbus_error_register_error_domain() in flatpak_error_quark(). It might need some decoding with g_dbus_error_strip_remote_error() tho (i never quite figured out when that is needed).

GDbusError documentation says that provided the g_dbus_error_register_error_domain is done on the receiving side, matching D-Bus errors will be turned back to their original domain and code.  g_dbus_error_strip_remote_error() is to remove the (wherever it's hidden?) D-Bus error name from the GError, so that it matches the originally-thrown GError (ie, it stops g_dbus_error_get_remote_error from working any more) - it's needed if you're presenting or propagating a GError which was returned to you by a method invocation. It's orthogonal to the reverse mapping of the D-Bus error to the GError domain/code, that's done based on the _register_error_domain lookup table during demarshalling - before the GError is handed to the application.

Long story short, if whatever is calling the system helper is checking for FLATPAK_ERROR_DOWNGRADE, what's here should work. :)

**Comment by mwleeds** [comment]
You saw the last fixup commit right?

File: system-helper/flatpak-system-helper.c [hunk 2]
```diff
@@ -355,8 +355,13 @@ handle_deploy (FlatpakSystemHelper   *object,
                                              ostree_progress,
                                              NULL, &error))
         {
-          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
-                                                 "Error pulling from repo: %s", error->message);
+          if (g_error_matches (error, FLATPAK_ERROR, FLATPAK_ERROR_DOWNGRADE))
+            g_dbus_method_invocation_return_error (invocation, FLATPAK_ERROR, FLATPAK_ERROR_DOWNGRADE,
+                                                   "%s", error->message);
+          else
```
[the remark this answers is not in the document]
Comment (mwleeds) [reply system-helper/flatpak-system-helper.c]: Sure, I didn't consider doing that. I rewrote it so see if it's good now

File: common/flatpak-utils.c [hunk 3]
```diff
@@ -52,6 +52,7 @@
 static const GDBusErrorEntry flatpak_error_entries[] = {
   {FLATPAK_ERROR_ALREADY_INSTALLED,     "org.freedesktop.Flatpak.Error.AlreadyInstalled"},
   {FLATPAK_ERROR_NOT_INSTALLED,         "org.freedesktop.Flatpak.Error.NotInstalled"},
+  {FLATPAK_ERROR_DOWNGRADE,             "org.freedesktop.Flatpak.Error.Downgrade"}, /* Since: 1.0 */
```
[the remark this answers is not in the document]
Comment (mwleeds) [reply common/flatpak-utils.c]: Added another commit for this

**Comment by alexlarsson** [comment]
@rh-atomic-bot r+
